### DISCUSSION TITLE
[General] Listening to L2 events using local-setup

### DISCUSSION THREAD

**Discussion by oac1771** [zkSync-Community-Hub/zksync-developers#110] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

Hello, I am trying to listen for Transfer events using a websocket connection to a [local instance of Zksync](https://github.com/matter-labs/local-setup) when minting an ERC721 token following [this guide](https://docs.zksync.io/dev/events/#watching-events).

### What is the specific issue or error you're encountering?

After minting a token, the websocket connection seems to make a call to the instance of Zksync, but returns an error instead of the actual transaction data.

### Can you share the error messages or logs you're receiving, if any?

When using the following websocket connection and filter I get the corresponding error:

```
const ws = new WebSocket('ws://localhost:3051');

setInterval(function () {
  ws.ping();
}, 10000);

ws.on('open', function open() {
  ws.send('{}');
});
```

error: `{"jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request"},"id":null}`

I also tried to validate that it fails correctly as stated [here](https://docs.zksync.io/dev/events/#filtering-events) by setting up an incorrect filter shown below and I still receive the same error message: `{"jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request"},"id":null}`

```
  const ws = new WebSocket('ws://localhost:3051');
  
  setInterval(function () {
    ws.ping();
  }, 10000);

  ws.on('open', function open() {
    ws.send('{"blocks": {"status": "committed"}}');
  });
```

### Have you made any recent changes to the contract before encountering this issue?

None

### Are there any external libraries or contracts that your contract interacts with?

None

### Can you provide the relevant portions of your contract code where the issue is occurring?

```
  const ws = new WebSocket('ws://localhost:3051');

  setInterval(function () {
    ws.ping();
  }, 10000);

  ws.on('open', function open() {
    ws.send('{}');
  });

  ws.on('close', function close(code, reason) {
    console.log(`Connection closed with code ${code}, reason: ${reason}`);
  });
  ws.on('message', function (data) {
    const event = JSON.parse(data);
    console.log(`data: ${data}`);
  });

   const tx = await contract.registerIdentity(); // this is the minting part
```

### Have you tried to isolate the problem, and if so, what were the results?

None

### What steps have you already taken to try to resolve the issue?

Attempted to validate that the connection fails as expected instead of returning the error I have been receiving

### Additional Details

It feels like there is something wrong with the structure of the requests being sent to my zksync instance when trying to get the event data but I dont know what additional setup I need to do on my end. 

Thanks!

**Comment by uF4No** [comment, marked as answer]
hey @oac1771 

Looks like you're mixing zksync Era with Lite.

zkSync Lite (previously known as zkSync v1.0) is the first version of the protocol and it's limited to token swaps and transactions. 
zkSync Era (previously known as zkSync v2.0) is the latest version and it's fully programmable and EVM compatible. You can deploy any smart contract.

The docs you were following are from zkSync Lite, whilst the in-memory node is a testing tool for zkSync Era. 

You can find how to handle events in zkSync Era here: https://era.zksync.io/docs/reference/architecture/events.html

**Reply by oac1771** [reply]
Ah ok my bad, thanks so much!

**Comment by threadaqged** [comment]
It seems like there might be an issue with the format of the requests being sent to your local zkSync instance via WebSocket. Double-check the format of the requests and ensure they comply with the expected format by the zkSync WebSocket server. Additionally, verify that the WebSocket server is properly configured to handle event subscription requests. If the issue persists, consider reaching out to the zkSync community or reviewing the WebSocket server's documentation for further assistance.

**Comment by bartlettdv** [comment]
Environment: Testnet
Issue: Invalid request error when trying to listen for Transfer events via websocket connection to local Zksync instance.
Error: {"jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid request"},"id":null}
Potential Solution: Ensure correct structure of requests sent to Zksync instance and validate websocket setup.

**Comment by tathimustingjason** [comment]
It seems there might be an issue with the structure of the WebSocket request being sent to the zkSync instance. Double-check the format of the request payload and ensure it aligns with the expected format for retrieving event data.

**Comment by spatounmas** [comment]
It seems that the websocket connection is encountering an "Invalid request" error when attempting to fetch event data from the local zkSync instance. This could be due to incorrect formatting of the request or missing parameters. Ensure that the websocket request is properly structured and includes all necessary parameters to retrieve the desired event data from zkSync. Double-check the documentation or specifications for the correct format of the websocket request to fetch Transfer events.
